### PR TITLE
Relationship between WebTransport, HTTP/3 and QUIC

### PR DESCRIPTION
**Pull request by LPardue**
Bernards commits
- Relationship between WebTransport, HTTP/3 and QUIC
- Update draft-ietf-webtrans-http3.md
- Update draft-ietf-webtrans-http3.md
- Update draft-ietf-webtrans-http3.md

My suggestions on top
  https://github.com/ietf-wg-webtrans/draft-ietf-webtrans-http3/pull/146/commits/05f31d55727611c35747146d10d7b3af93b20898

### CODE REVIEW

File: draft-ietf-webtrans-http3.md [hunk 1]
```diff
@@ -76,7 +76,55 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+"QUIC: A UDP-Based Multiplexed and Secure Transport" {{!RFC9000}}
+defines QUIC security, stream multiplexing, reliable and ordered streams,
+stream HoL blocking avoidance, flow control, and congestion control.
+The transport layer provides these services to applications but does
+not constrain how streams are used.
+
+HTTP is an application-layer protocol, defined by "HTTP Semantics" {{!RFC9110}}.
+HTTP/3 has specific features that are not part of the HTTP semantics: QPACK
+header compression (static and dynamic) and Server Push. Of these, only
+static decompression is mandatory to support.
```
Comment (afrind) [inline draft-ietf-webtrans-http3.md]: QPACK static header field decompression

File: draft-ietf-webtrans-http3.md [hunk 2]
```diff
@@ -76,7 +76,55 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+
+WebTransport over HTTP/3 makes it possible for an application to directly access
+QUIC transport via an HTTP/3 connection. HTTP/3 is the application mapping for
+QUIC, defined in {{!RFC9113}}, which describes how streams are used to carry control
+data or HTTP request and response message sequences in the form of frames, as
+well as describing details of stream and connection lifecycle management.
+
+However, although WebTransport requires HTTP for its handshake, it doesn't require
```
Comment (afrind) [inline draft-ietf-webtrans-http3.md]: However and although are redundant?  Just "Although WebTransport..." ?

File: draft-ietf-webtrans-http3.md (hunk 1 above)
Comment (afrind) [reply draft-ietf-webtrans-http3.md]: And huffman literal decoding?  @vasilvv mentioned this was the most annoying part, wondering if we can create a profile or setting opting out of huffman support.

File: draft-ietf-webtrans-http3.md [hunk 3]
```diff
@@ -76,7 +76,55 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+HTTP for any other part. This makes it possible to create a minimal WebTransport client
+or server that supports only the semantics required to complete the handshake.
+This involves generating/parsing the request method, host,
+path, protocol, optional Origin header, and perhaps some extra headers;
+generating/parsing the response status code, and possibly some extra headers.
+The receiver can likely perform checks using bytestring comparisons.
+
+Rather than having to interact with QUIC streams via HTTP semantics and
+HTTP/3 frames, a WebTransport session allows direct access. This relies on
```
Comment (afrind) [inline draft-ietf-webtrans-http3.md]: maybe "interact with QUIC features" and "allows direct access to them"

File: draft-ietf-webtrans-http3.md [hunk 4]
```diff
@@ -76,7 +76,55 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+the WebTransport handshake (extended CONNECT method) to provide some prior checks.
+This is important or the Web security model where same-origin and cross-origin resource
```
Comment (afrind) [inline draft-ietf-webtrans-http3.md]: s/or/for/?

**Review by martinthomson — COMMENTED** [review]
See my feedback on #145.   Most of it applies here too.

File: draft-ietf-webtrans-http3.md [hunk 5]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
-# Protocol Overview
+# Overview
+
+## QUIC, WebTransport, and HTTP/3
+
+QUIC version 1 {{!RFC9000}} is a secure transport protocol with flow control and
+congestion control. QUIC supports application data exchange via streams;
+reliable and ordered byte streams that can be multiplexed. Stream independence
+can mitigate head-of-line blocking. While QUIC provides streams as a transport
+service, it is unopinionated about their usageage. Applicability of streams is
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:87]:
```suggestion
service, it is unopinionated about their usage. The applicability of streams is
```

File: draft-ietf-webtrans-http3.md [hunk 6]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+
+HTTP is an application-layer protocol, defined by "HTTP Semantics" {{!RFC9110}}.
+HTTP/3 is the application mapping for QUIC, defined in {{!RFC9113}}. It
+describes how streams are used to carry control data or HTTP request and
+response message sequences in the form of frames, as well as describing details
+of stream and connection lifecycle management.
+
+WebTransport over HTTP/3 makes it possible for an application to directly access
+QUIC transport via an HTTP/3 connection. This relies on the WebTransport
+handshake (extended CONNECT method) to provide some prior checks. This is
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:98]:
```suggestion
handshake, which uses the extended CONNECT method, to perform important checks. This is
```

File: draft-ietf-webtrans-http3.md [hunk 7]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+important or the Web security model where same-origin and cross-origin resource
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:99]:
```suggestion
necessary for the Web security model where same-origin and cross-origin resource
```

File: draft-ietf-webtrans-http3.md [hunk 8]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+access is very important. Post-handshake, QUIC streams use header bytes for
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:100]:
```suggestion
access may not be the same. Post-handshake, QUIC streams use header bytes for
```
Or some better way to say this

File: draft-ietf-webtrans-http3.md [hunk 9]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+accounting purposes, but after that an application can use QUIC streams however
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:101]: I'm not sure I totally follow this sentence. What is it trying to say?
Is this trying to say that WebTransport uses HTTP/3 to convey header bytes via a QUIC stream, but then eventually just uses the QUIC stream directly?

File: draft-ietf-webtrans-http3.md [hunk 10]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+it would like. This is similar to WebSockets over HTTP/1.1, where access is
+enabled to the underlying bytestream after both sides have agreed the handshake.
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:103]:
```suggestion
enabled to the underlying bytestream after both sides have completed the handshake.
```

File: draft-ietf-webtrans-http3.md [hunk 11]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+
+~~~~~~~~~~ drawing
+|       WebTransport       |
+|  Punch  | HTTP Semantics |
+| Through |   HTTP/3       |
+|          QUIC            |
+~~~~~~~~~~
+{: #fig-webtransport-layers title="WebTransport Layering"}
+
+While WebTransport requires HTTP for its handshake, it doesn't require HTTP for
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:114]:
```suggestion
While WebTransport requires HTTP for the handshake, it doesn't require HTTP for
```

File: draft-ietf-webtrans-http3.md [hunk 12]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+any other part. HTTP/3 offers two features in addition to HTTP Semantics: QPACK
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:115]:
```suggestion
anything else. HTTP/3 offers two features in addition to HTTP Semantics: QPACK
```

**Review by ekinnear — APPROVED** [review]

**Comment by DavidSchinazi** [comment]
Chair: next step is for @LPardue to do an editorial pass based on received comments

File: draft-ietf-webtrans-http3.md (hunk 12 above)
Comment (aboba) [reply draft-ietf-webtrans-http3.md]: fixed.

File: draft-ietf-webtrans-http3.md (hunk 4 above)
Comment (aboba) [reply draft-ietf-webtrans-http3.md]: fixed.

File: draft-ietf-webtrans-http3.md (hunk 3 above)
Comment (aboba) [reply draft-ietf-webtrans-http3.md]: It now says "WebTransport over HTTP/3 makes it possible for an application to directly access
QUIC transport via an HTTP/3 connection.".

**Comment by aboba** [comment]
@DavidSchinazi  I have gone ahead and merged several of the proposed changes, and where changes were made in response to comments have resolved the conversations.  There are still a few outstanding comments.

File: draft-ietf-webtrans-http3.md (hunk 1 above)
Comment (aboba) [reply draft-ietf-webtrans-http3.md]: These comments don't appear to have been incorporated. The text now says: "While WebTransport requires HTTP for the handshake, it doesn't require HTTP for anything else. HTTP/3 offers two features in addition to HTTP Semantics: QPACK header compression and Server Push. Neither of these can be used to the benefit
of WebTransport after the handshake succeeds."

File: draft-ietf-webtrans-http3.md (hunk 9 above)
Comment (aboba) [reply draft-ietf-webtrans-http3.md:101]: Maybe it's trying to say that a WebTransport application has access to QUIC streams, other than the header bytes?

**Comment by LPardue** [comment]
Apologies for the delay here, I will pick this up again this week.

File: draft-ietf-webtrans-http3.md [hunk 13]
```diff
@@ -76,7 +76,62 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+it would like. This is similar to WebSockets over HTTP/1.1, where access is
+enabled to the underlying bytestream after both sides have completed the handshake.
+As a result, WebTransport layering appears as follows:
+
+~~~~~~~~~~ drawing
+|       WebTransport       |
+|  Punch  | HTTP Semantics |
+| Through |   HTTP/3       |
+|          QUIC            |
+~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (LPardue) [reply draft-ietf-webtrans-http3.md:111]: punchout

**Comment by LPardue** [comment]
So picking this up was harder than I imagined because I had to address comments from Martin in a different PR that addressed text that had since had some work done on them. I've tried my best to incorporated the suggestions made there, and to address Eric and Alan's questions. As a result, I needed to hammer on the text even more. But I like the result.

The one thing unaddressed is @afrind asked "wondering if we can create a profile or setting opting out of huffman support.". I don't want to take on that task in this PR. If the WG wants to bring that up, then a new issue is the best place for further discussion. A subsequent proposal to introduce such an extension would then need to make it's own PR to update this document accordingly.

File: draft-ietf-webtrans-http3.md [hunk 14]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+congestion control. QUIC supports application data exchange via streams;
+reliable and ordered byte streams that can be multiplexed. Stream independence
+can mitigate head-of-line blocking. While QUIC provides streams as a transport
+service, it is unopinionated about their usage. The applicability of streams is
+described by section 4 of {{?RFC9308}}.
+
+HTTP is an application-layer protocol, defined by "HTTP Semantics" {{!RFC9110}}.
+HTTP/3 is the application mapping for QUIC, defined in {{!RFC9114}}. It
+describes how QUIC streams are used to carry control data or HTTP request and
+response message sequences in the form of frames, as well as describing details
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:93]:
```suggestion
response message sequences in the form of frames and describes details
```

File: draft-ietf-webtrans-http3.md [hunk 15]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+addition to HTTP Semantics: QPACK header compression {{?RFC9208}} and Server
+Push {{Section 4.6 of RFC9114}}.
+
+WebTransport session establishment involves interacting at the HTTP layer with a
+resource. For Web user agents, this interaction is important for security
+reasons, especially to ensure that the resource is willing to use WebTransport.
+
+Although WebTransport requires HTTP for its handshake, when HTTP/3 is in use,
+HTTP is not used for anything else related to an established session. Instead,
+QUIC streams begin a header sequence of bytes that links them to the established
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:104]: If I'm understanding correctly:
```suggestion
QUIC streams begin with a header sequence of bytes that links them to the established
```

File: draft-ietf-webtrans-http3.md [hunk 16]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+session. The remainder of the stream is the body, which carries the payload of
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:105]:
```suggestion
session. The remainder of the stream is the body, which carries the payload specified by
```
or
```suggestion
session. The remainder of the stream is the body, which carries the payload supplied by
```

File: draft-ietf-webtrans-http3.md [hunk 17]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+the application using WebTransport. This process is similar to WebSockets over
+HTTP/1.1 {{?RFC6455}}, where access  to the underlying bytestream is enabled
```
Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:107]:
```suggestion
HTTP/1.1 {{?RFC6455}}, where access to the underlying bytestream is enabled
```

File: draft-ietf-webtrans-http3.md [hunk 18]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+{: #fig-webtransport-layers title="WebTransport Layering"}
+
+### Minimizing Implementation Complexity
+
+WebTransport has minimal interaction with HTTP and HTTP/3. Clients or servers
+can constrain their use of features to only those required to complete a
+WebTransport handshake:
+
+* Generating/parsing the request method, host, path, protocol, optional Origin
+header, and perhaps some extra headers
```
Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:134]:
```suggestion
header, and perhaps some extra headers.
```

File: draft-ietf-webtrans-http3.md (hunk 17 above)
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md]:
```suggestion
HTTP/1.1 {{?RFC6455}}, where access to the underlying bytestream is enabled
```

File: draft-ietf-webtrans-http3.md [hunk 19]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+* Generating/parsing the response status code, and possibly some extra headers.
+
+The receiver can likely perform several of its HTTP-level requirements using
+bytestring comparisons.
+
+While HTTP/3 encodes HTTP messages using QPACK, the complexity can be minimized.
+Receivers can disable dynamic decompression entirely but must always support
+static decompression and Huffman decoding. Senders can opt to never use dyanamic
+compression, static compression, or Huffman encoding. support and to minimize
+their use of  this can be constrained by disabling dynamic compression
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:145]:
```suggestion
compression, static compression, or Huffman encoding.
```

Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:145]:
```suggestion
compression, static compression, or Huffman encoding.
```

File: draft-ietf-webtrans-http3.md [hunk 20]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+
+Although WebTransport requires HTTP for its handshake, when HTTP/3 is in use,
+HTTP is not used for anything else related to an established session. Instead,
+QUIC streams begin a header sequence of bytes that links them to the established
+session. The remainder of the stream is the body, which carries the payload of
+the application using WebTransport. This process is similar to WebSockets over
+HTTP/1.1 {{?RFC6455}}, where access  to the underlying bytestream is enabled
+after both sides have completed the handshake.
+
+The layering of QUIC, HTTP/3 and WebTransport is shown in
```
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:110]:
```suggestion
The layering of QUIC, HTTP/3, and WebTransport is shown in
```

File: draft-ietf-webtrans-http3.md [hunk 21]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+header, and perhaps some extra headers
+
+* Generating/parsing the response status code, and possibly some extra headers.
+
+The receiver can likely perform several of its HTTP-level requirements using
+bytestring comparisons.
+
+While HTTP/3 encodes HTTP messages using QPACK, the complexity can be minimized.
+Receivers can disable dynamic decompression entirely but must always support
+static decompression and Huffman decoding. Senders can opt to never use dyanamic
```
Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:143]:
```suggestion
static decompression and Huffman decoding. Senders can opt to never use dynamic
```

File: draft-ietf-webtrans-http3.md (hunk 18 above)
Comment (ekinnear) [inline draft-ietf-webtrans-http3.md:134]:
```suggestion
header, and perhaps some extra headers.
```

**Review by ekinnear — APPROVED** [review]

File: draft-ietf-webtrans-http3.md [hunk 22]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+WebTransport session establishment involves interacting at the HTTP layer with a
+resource. For Web user agents, this interaction is important for security
+reasons, especially to ensure that the resource is willing to use WebTransport.
+
+Although WebTransport requires HTTP for its handshake, when HTTP/3 is in use,
+HTTP is not used for anything else related to an established session. Instead,
+QUIC streams begin a header sequence of bytes that links them to the established
+session. The remainder of the stream is the body, which carries the payload of
+the application using WebTransport. This process is similar to WebSockets over
+HTTP/1.1 {{?RFC6455}}, where access to the underlying bytestream is enabled
```
Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:107]:
```suggestion
HTTP/1.1 {{?ORIGIN=RFC6455}}, where access to the underlying byte stream is enabled
```

File: draft-ietf-webtrans-http3.md [hunk 23]
```diff
@@ -76,7 +76,75 @@ server.  An HTTP/3 server is the server that terminates HTTP/3 connections; a
...
+The layering of QUIC, HTTP/3 and WebTransport is shown in
+{{fig-webtransport-layers}}. Once a WebTransport session is established,
+applications have nearly direct access to QUIC.
+
+~~~~~~~~~~ aasvg
+,--------------------------------,
+|            WebTransport        |
+,----------------,---------------,
+| HTTP Semantics |               |
+,----------------, Nearly direct |
```
Comment (martinthomson) [inline draft-ietf-webtrans-http3.md:119]:
```suggestion
| HTTP Semantics |               |
|       and      |               |
| Session Setup | Nearly direct |
,----------------,               |
```

**Comment by LPardue** [comment]
I authorize any author to fix the nits, please make the pain stop.

**Comment by DavidSchinazi** [comment]
Sorry, the GitHub UI completely fell over on me

**Comment by aboba** [comment]
Yeah!

**Comment by LPardue** [comment]
Thanks team!

**Pull request merged**